sklearn/decomposition/tests/test_nmf.py

```diff
@@ -1060,3 +1060,44 @@ def test_nmf_custom_init_shape_error():
 
     with pytest.raises(ValueError, match="Array with wrong second dimension passed"):
         nmf.fit(X, H=H, W=rng.random_sample((6, 3)))
+
+
+def test_nmf_objective_function(global_random_seed):
+    """Test that the objective function is correctly computed."""
+    rng = np.random.RandomState(global_random_seed)
+
+    n_samples = 100
+    n_features = 10
+    alpha_W = 0.1
+    alpha_H = 0.2
+    l1_ratio = 0.5
+    beta_loss = 1
+
+    X = rng.random_sample((n_samples, n_features))
+
+    model = NMF(
+        n_components=5,
+        beta_loss=beta_loss,
+        alpha_W=alpha_W,
+        alpha_H=alpha_H,
+        l1_ratio=l1_ratio,
+        solver="mu",
+        random_state=global_random_seed,
+    ).fit(X)
+
+    W, H = model.transform(X), model.components_
+
+    obj_func = model.objective_function(X, normalize=False)
+
+    expected_data_fit = nmf._beta_divergence(X, W, H, beta_loss)
+    expected_penalty = (
+        alpha_W * l1_ratio * n_features * W.sum()
+        + alpha_H * l1_ratio * n_samples * H.sum()
+        + 0.5 * alpha_W * (1 - l1_ratio) * n_features * (W**2).sum()
+        + 0.5 * alpha_H * (1 - l1_ratio) * n_samples * (H**2).sum()
+    )
```

Comment:
Ideally, one might also want to check that the objective function is actually monotonically decreasing at each iteration but this kind of tests would only be easy to write once we have the callbacks API.

Reply:
We'll be able to add such tests easily when we have the callbacks